File: pylib/tools/markpure.py
# Copyright: Ankitects Pty Ltd and contributors
# License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

import os
import re
import sys

root = sys.argv[1]

type_re = re.compile(r'(make(Enum|MessageType))\(\n\s+".*",')
for dirpath, dirnames, filenames in os.walk(root):
    for filename in filenames:
        if filename.endswith(".js"):
            file = os.path.join(dirpath, filename)
            with open(file, "r", encoding="utf8") as f:
                contents = f.read()

            # allow tree shaking on proto messages
            contents = contents.replace(
                "= proto3.make", "= /* @__PURE__ */ proto3.make"
            )
            # strip out typeName info, which appears to only be required for
            # certain JSON functionality (though this only saves a few hundred
            # bytes)
            contents = type_re.sub('\\1("",', contents)

            with open(file, "w", encoding="utf8") as f:
                f.write(contents)

File: ts/tools/markpure.ts
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

import * as fs from "fs";
import * as path from "path";

function allFilesInDir(directory): string[] {
    let results: string[] = [];
    const list = fs.readdirSync(directory);

    list.forEach(function(file) {
        file = path.join(directory, file);
        const stat = fs.statSync(file);

        if (stat && stat.isDirectory()) {
            results = results.concat(allFilesInDir(file));
        } else {
            results.push(file);
        }
    });

    return results;
}

function adjustFiles() {
    const root = process.argv[2];
    const typeRe = /(make(Enum|MessageType))\(\n\s+".*",/g;

    const jsFiles = allFilesInDir(root).filter(f => f.endsWith(".js"));
    for (const file of jsFiles) {
        let contents = fs.readFileSync(file, "utf8");

        // allow tree shaking on proto messages
        contents = contents.replace(
            /= proto3.make/g,
            "= /* @__PURE__ */ proto3.make",
        );

        // strip out typeName info, which appears to only be required for
        // certain JSON functionality (though this only saves a few hundred
        // bytes)
        contents = contents.replace(typeRe, "$1(\"\",");

        fs.writeFileSync(file, contents, "utf8");
    }
}

adjustFiles();
